### ISSUE TITLE
Try enabling Scotch TPL in SEMS build using sems-scotch/6.0.3/nopthread_64bit_parallel

### ISSUE THREAD

**Issue by bartlettroscoe** [trilinos/Trilinos#2051]
CC: @trilinos/framework, @trilinos/zoltan, @trilinos/zoltan2, @trilinos/shylu 

## Motivation and Context

Scotch is an important TPL for Trilinos and is supported by several Trilinos packages and is used by some Trilinos customers (see discussions in #410).

This story is to try to enable the Scotch TPL in the standard CI build of Trilinos using the SEMS module `sems-scotch/6.0.3/nopthread_64bit_parallel`. 

## Definition of Done

* Documentation results of experiment to enable Scotch using `sems-scotch/6.0.3/nopthread_64bit_parallel` for standard CI build of Trilinos (see #482, #1304)
* If all tests pass, then update `develop` branch with Scotch enabled.

## Related Issues

* Follows: #475, #476, #482
* Related to: #1400, #1304

**Comment by bartlettroscoe** [comment]
All of the Zoltan tests pass after enabling this Scotch TPL in commit b20b93c.  The results from testing are shown below.


In particular it showed:

```
52/68 Test  #4: Zoltan_ch_brack2_3_scotch_parallel .........   Passed    0.05 sec
53/68 Test  #7: Zoltan_ch_bug_scotch_parallel ..............   Passed    0.05 sec
54/68 Test #10: Zoltan_ch_degenerate_scotch_parallel .......   Passed    0.05 sec
55/68 Test #13: Zoltan_ch_degenerateAA_scotch_parallel .....   Passed    0.05 sec
56/68 Test #16: Zoltan_ch_drake_scotch_parallel ............   Passed    0.05 sec
57/68 Test #19: Zoltan_ch_ewgt_scotch_parallel .............   Passed    1.12 sec
58/68 Test #22: Zoltan_ch_grid20x19_scotch_parallel ........   Passed    0.05 sec
59/68 Test #25: Zoltan_ch_hammond_scotch_parallel ..........   Passed    1.15 sec
60/68 Test #28: Zoltan_ch_hammond2_scotch_parallel .........   Passed    0.05 sec
61/68 Test #31: Zoltan_ch_hughes_scotch_parallel ...........   Passed    0.05 sec
62/68 Test #34: Zoltan_ch_nograph_scotch_parallel ..........   Passed    0.05 sec
63/68 Test #37: Zoltan_ch_onedbug_scotch_parallel ..........   Passed    0.05 sec
64/68 Test #40: Zoltan_ch_serial_scotch_parallel ...........   Passed    0.05 sec
65/68 Test #43: Zoltan_ch_simple_scotch_parallel ...........   Passed    2.23 sec
66/68 Test #46: Zoltan_ch_simple3d_scotch_parallel .........   Passed    0.05 sec
67/68 Test #49: Zoltan_ch_vwgt_scotch_parallel .............   Passed    1.10 sec
68/68 Test #52: Zoltan_ch_vwgt2_scotch_parallel ............   Passed    0.05 sec

...

100% tests passed, 0 tests failed out of 68

Label Time Summary:
Zoltan    = 1176.31 sec (68 tests)

Total Test time (real) = 511.49 sec
```

Now I am running the full CI build and test suite with:

```
$ ./checkin-test-sems.sh --enable-all-packages=on --local-do-all
```

<details>
<summary><b>Detailed Zoltan test results</b> (click to expend)</b></summary>

```
Test project /ascldap/users/rabartl/Trilinos.base/BUILDS/GCC-4.8.4/MPI_RELEASE_DEBUG_SHARED_PT
      Start 59: Zoltan_hg_simple_zoltan_parallel
      Start 57: Zoltan_hg_ibm03_zoltan_parallel
      Start 47: Zoltan_ch_vwgt_zoltan_parallel
 1/68 Test #47: Zoltan_ch_vwgt_zoltan_parallel .............   Passed   53.79 sec
      Start  5: Zoltan_ch_bug_zoltan_parallel
 2/68 Test #57: Zoltan_hg_ibm03_zoltan_parallel ............   Passed   78.29 sec
      Start 23: Zoltan_ch_hammond_zoltan_parallel
      Start 38: Zoltan_ch_serial_zoltan_parallel
 3/68 Test #38: Zoltan_ch_serial_zoltan_parallel ...........   Passed    0.40 sec
 4/68 Test #59: Zoltan_hg_simple_zoltan_parallel ...........   Passed   89.63 sec
      Start 60: Zoltan_hg_vwgt_zoltan_parallel
 5/68 Test  #5: Zoltan_ch_bug_zoltan_parallel ..............   Passed   44.73 sec
      Start 41: Zoltan_ch_simple_zoltan_parallel
 6/68 Test #23: Zoltan_ch_hammond_zoltan_parallel ..........   Passed   77.38 sec
      Start 55: Zoltan_hg_ewgt_zoltan_parallel
      Start 17: Zoltan_ch_ewgt_zoltan_parallel
 7/68 Test #60: Zoltan_hg_vwgt_zoltan_parallel .............   Passed   68.16 sec
      Start 54: Zoltan_hg_diag500_4_zoltan_parallel
 8/68 Test #41: Zoltan_ch_simple_zoltan_parallel ...........   Passed   67.26 sec
      Start 58: Zoltan_hg_nograph_zoltan_parallel
 9/68 Test #17: Zoltan_ch_ewgt_zoltan_parallel .............   Passed   44.04 sec
      Start 56: Zoltan_hg_felix_zoltan_parallel
10/68 Test #54: Zoltan_hg_diag500_4_zoltan_parallel ........   Passed   43.13 sec
      Start 35: Zoltan_ch_onedbug_zoltan_parallel
11/68 Test #58: Zoltan_hg_nograph_zoltan_parallel ..........   Passed   43.23 sec
      Start 32: Zoltan_ch_nograph_zoltan_parallel
12/68 Test #55: Zoltan_hg_ewgt_zoltan_parallel .............   Passed   65.90 sec
      Start 50: Zoltan_ch_vwgt2_zoltan_parallel
13/68 Test #50: Zoltan_ch_vwgt2_zoltan_parallel ............   Passed   15.03 sec
      Start 14: Zoltan_ch_drake_zoltan_parallel
14/68 Test #35: Zoltan_ch_onedbug_zoltan_parallel ..........   Passed   41.98 sec
      Start 67: Zoltan_problemGRAPH_MPI_4
15/68 Test #14: Zoltan_ch_drake_zoltan_parallel ............   Passed    6.83 sec
16/68 Test #67: Zoltan_problemGRAPH_MPI_4 ..................   Passed    1.21 sec
      Start 20: Zoltan_ch_grid20x19_zoltan_parallel
17/68 Test #32: Zoltan_ch_nograph_zoltan_parallel ..........   Passed   41.95 sec
      Start 26: Zoltan_ch_hammond2_zoltan_parallel
18/68 Test #56: Zoltan_hg_felix_zoltan_parallel ............   Passed   66.32 sec
      Start  2: Zoltan_ch_brack2_3_zoltan_parallel
19/68 Test  #2: Zoltan_ch_brack2_3_zoltan_parallel .........   Passed    8.42 sec
      Start 64: Zoltan_simpleGRAPH_MPI_4
20/68 Test #64: Zoltan_simpleGRAPH_MPI_4 ...................   Passed    1.20 sec
      Start  1: Zoltan_test_get_callbacks_MPI_4
21/68 Test  #1: Zoltan_test_get_callbacks_MPI_4 ............   Passed    1.20 sec
      Start 62: Zoltan_simpleBLOCK_MPI_4
22/68 Test #26: Zoltan_ch_hammond2_zoltan_parallel .........   Passed   26.92 sec
      Start 11: Zoltan_ch_degenerateAA_zoltan_parallel
23/68 Test #62: Zoltan_simpleBLOCK_MPI_4 ...................   Passed    1.25 sec
      Start 65: Zoltan_simplePHG_MPI_4
24/68 Test #65: Zoltan_simplePHG_MPI_4 .....................   Passed    1.10 sec
      Start 63: Zoltan_simpleRCB_MPI_4
25/68 Test #63: Zoltan_simpleRCB_MPI_4 .....................   Passed    1.10 sec
      Start 61: Zoltan_simpleHIER_MPI_4
26/68 Test #61: Zoltan_simpleHIER_MPI_4 ....................   Passed    1.10 sec
      Start 66: Zoltan_migrateGRAPH_MPI_4
27/68 Test #66: Zoltan_migrateGRAPH_MPI_4 ..................   Passed    1.10 sec
      Start 68: Zoltan_exampleBLOCK_MPI_4
28/68 Test #68: Zoltan_exampleBLOCK_MPI_4 ..................   Passed    1.10 sec
      Start 44: Zoltan_ch_simple3d_zoltan_parallel
29/68 Test #44: Zoltan_ch_simple3d_zoltan_parallel .........   Passed    0.20 sec
30/68 Test #11: Zoltan_ch_degenerateAA_zoltan_parallel .....   Passed    6.60 sec
      Start 53: Zoltan_hg_cage10_zoltan_parallel
31/68 Test #20: Zoltan_ch_grid20x19_zoltan_parallel ........   Passed   46.34 sec
32/68 Test #53: Zoltan_hg_cage10_zoltan_parallel ...........   Passed   54.46 sec
      Start 42: Zoltan_ch_simple_parmetis_parallel
33/68 Test #42: Zoltan_ch_simple_parmetis_parallel .........   Passed   27.77 sec
      Start 24: Zoltan_ch_hammond_parmetis_parallel
34/68 Test #24: Zoltan_ch_hammond_parmetis_parallel ........   Passed   26.97 sec
      Start 48: Zoltan_ch_vwgt_parmetis_parallel
35/68 Test #48: Zoltan_ch_vwgt_parmetis_parallel ...........   Passed   23.40 sec
      Start 18: Zoltan_ch_ewgt_parmetis_parallel
36/68 Test #18: Zoltan_ch_ewgt_parmetis_parallel ...........   Passed   22.65 sec
      Start 21: Zoltan_ch_grid20x19_parmetis_parallel
37/68 Test #21: Zoltan_ch_grid20x19_parmetis_parallel ......   Passed   18.28 sec
      Start  6: Zoltan_ch_bug_parmetis_parallel
38/68 Test  #6: Zoltan_ch_bug_parmetis_parallel ............   Passed   16.05 sec
      Start 33: Zoltan_ch_nograph_parmetis_parallel
39/68 Test #33: Zoltan_ch_nograph_parmetis_parallel ........   Passed    9.71 sec
      Start  3: Zoltan_ch_brack2_3_parmetis_parallel
40/68 Test  #3: Zoltan_ch_brack2_3_parmetis_parallel .......   Passed    6.85 sec
      Start  8: Zoltan_ch_degenerate_zoltan_parallel
      Start 29: Zoltan_ch_hughes_zoltan_parallel
41/68 Test #29: Zoltan_ch_hughes_zoltan_parallel ...........   Passed    3.47 sec
42/68 Test  #8: Zoltan_ch_degenerate_zoltan_parallel .......   Passed    6.65 sec
      Start 51: Zoltan_ch_vwgt2_parmetis_parallel
43/68 Test #51: Zoltan_ch_vwgt2_parmetis_parallel ..........   Passed    4.32 sec
      Start 36: Zoltan_ch_onedbug_parmetis_parallel
44/68 Test #36: Zoltan_ch_onedbug_parmetis_parallel ........   Passed    2.22 sec
      Start  9: Zoltan_ch_degenerate_parmetis_parallel
45/68 Test  #9: Zoltan_ch_degenerate_parmetis_parallel .....   Passed    0.07 sec
      Start 45: Zoltan_ch_simple3d_parmetis_parallel
46/68 Test #45: Zoltan_ch_simple3d_parmetis_parallel .......   Passed    0.05 sec
      Start 39: Zoltan_ch_serial_parmetis_parallel
47/68 Test #39: Zoltan_ch_serial_parmetis_parallel .........   Passed    0.05 sec
      Start 30: Zoltan_ch_hughes_parmetis_parallel
48/68 Test #30: Zoltan_ch_hughes_parmetis_parallel .........   Passed    0.05 sec
      Start 12: Zoltan_ch_degenerateAA_parmetis_parallel
49/68 Test #12: Zoltan_ch_degenerateAA_parmetis_parallel ...   Passed    0.05 sec
      Start 27: Zoltan_ch_hammond2_parmetis_parallel
50/68 Test #27: Zoltan_ch_hammond2_parmetis_parallel .......   Passed    0.05 sec
      Start 15: Zoltan_ch_drake_parmetis_parallel
51/68 Test #15: Zoltan_ch_drake_parmetis_parallel ..........   Passed    0.05 sec
      Start  4: Zoltan_ch_brack2_3_scotch_parallel
52/68 Test  #4: Zoltan_ch_brack2_3_scotch_parallel .........   Passed    0.05 sec
      Start  7: Zoltan_ch_bug_scotch_parallel
53/68 Test  #7: Zoltan_ch_bug_scotch_parallel ..............   Passed    0.05 sec
      Start 10: Zoltan_ch_degenerate_scotch_parallel
54/68 Test #10: Zoltan_ch_degenerate_scotch_parallel .......   Passed    0.05 sec
      Start 13: Zoltan_ch_degenerateAA_scotch_parallel
55/68 Test #13: Zoltan_ch_degenerateAA_scotch_parallel .....   Passed    0.05 sec
      Start 16: Zoltan_ch_drake_scotch_parallel
56/68 Test #16: Zoltan_ch_drake_scotch_parallel ............   Passed    0.05 sec
      Start 19: Zoltan_ch_ewgt_scotch_parallel
57/68 Test #19: Zoltan_ch_ewgt_scotch_parallel .............   Passed    1.12 sec
      Start 22: Zoltan_ch_grid20x19_scotch_parallel
58/68 Test #22: Zoltan_ch_grid20x19_scotch_parallel ........   Passed    0.05 sec
      Start 25: Zoltan_ch_hammond_scotch_parallel
59/68 Test #25: Zoltan_ch_hammond_scotch_parallel ..........   Passed    1.15 sec
      Start 28: Zoltan_ch_hammond2_scotch_parallel
60/68 Test #28: Zoltan_ch_hammond2_scotch_parallel .........   Passed    0.05 sec
      Start 31: Zoltan_ch_hughes_scotch_parallel
61/68 Test #31: Zoltan_ch_hughes_scotch_parallel ...........   Passed    0.05 sec
      Start 34: Zoltan_ch_nograph_scotch_parallel
62/68 Test #34: Zoltan_ch_nograph_scotch_parallel ..........   Passed    0.05 sec
      Start 37: Zoltan_ch_onedbug_scotch_parallel
63/68 Test #37: Zoltan_ch_onedbug_scotch_parallel ..........   Passed    0.05 sec
      Start 40: Zoltan_ch_serial_scotch_parallel
64/68 Test #40: Zoltan_ch_serial_scotch_parallel ...........   Passed    0.05 sec
      Start 43: Zoltan_ch_simple_scotch_parallel
65/68 Test #43: Zoltan_ch_simple_scotch_parallel ...........   Passed    2.23 sec
      Start 46: Zoltan_ch_simple3d_scotch_parallel
66/68 Test #46: Zoltan_ch_simple3d_scotch_parallel .........   Passed    0.05 sec
      Start 49: Zoltan_ch_vwgt_scotch_parallel
67/68 Test #49: Zoltan_ch_vwgt_scotch_parallel .............   Passed    1.10 sec
      Start 52: Zoltan_ch_vwgt2_scotch_parallel
68/68 Test #52: Zoltan_ch_vwgt2_scotch_parallel ............   Passed    0.05 sec

100% tests passed, 0 tests failed out of 68

Label Time Summary:
Zoltan    = 1176.31 sec (68 tests)

Total Test time (real) = 511.49 sec
```
</details>

**Comment by bartlettroscoe** [comment]
Running:

```
$ ./checkin-test-sems.sh --enable-all-packages=on --local-do-all
```

on the commit b20b93c I got all passing tests except for one shown in the test results:

```
  99% tests passed, 1 tests failed out of 2437
  
  Label Time Summary:
  Amesos               =  19.47 sec (14 tests)
  Amesos2              =  10.97 sec (8 tests)
  Anasazi              = 103.20 sec (71 tests)
  AztecOO              =  16.20 sec (17 tests)
  Belos                =  99.78 sec (70 tests)
  Domi                 = 149.25 sec (125 tests)
  Epetra               =  46.66 sec (61 tests)
  EpetraExt            =  14.04 sec (10 tests)
  FEI                  =  46.37 sec (43 tests)
  Galeri               =   4.41 sec (9 tests)
  GlobiPack            =   2.33 sec (6 tests)
  Ifpack               =  59.80 sec (53 tests)
  Ifpack2              =  41.68 sec (35 tests)
  Intrepid             = 192.36 sec (152 tests)
  Intrepid2            =  50.82 sec (90 tests)
  Isorropia            =   8.12 sec (6 tests)
  Kokkos               =  33.84 sec (21 tests)
  KokkosKernels        =  17.65 sec (4 tests)
  ML                   =  47.26 sec (34 tests)
  MiniTensor           =   0.94 sec (2 tests)
  MueLu                = 283.18 sec (79 tests)
  NOX                  = 144.43 sec (106 tests)
  OptiPack             =   6.37 sec (5 tests)
  Panzer               = 551.09 sec (147 tests)
  Phalanx              =   8.98 sec (24 tests)
  Pike                 =   4.29 sec (7 tests)
  Piro                 =  25.52 sec (12 tests)
  ROL                  = 758.22 sec (145 tests)
  RTOp                 =  11.18 sec (24 tests)
  Rythmos              = 157.20 sec (83 tests)
  SEACAS               =  14.17 sec (14 tests)
  STK                  =  27.39 sec (12 tests)
  Sacado               =  74.38 sec (292 tests)
  Shards               =   1.11 sec (4 tests)
  Stokhos              =  97.23 sec (75 tests)
  Stratimikos          =  35.74 sec (39 tests)
  Teko                 =  74.90 sec (19 tests)
  Tempus               = 992.58 sec (11 tests)
  Teuchos              =  56.72 sec (128 tests)
  ThreadPool           =   7.62 sec (10 tests)
  Thyra                =  60.99 sec (80 tests)
  Tpetra               = 146.40 sec (138 tests)
  TrilinosCouplings    =  42.04 sec (19 tests)
  Triutils             =   2.42 sec (2 tests)
  Xpetra               =  44.88 sec (18 tests)
  Zoltan               = 218.89 sec (19 tests)
  Zoltan2              = 140.83 sec (102 tests)
  
  Total Test time (real) = 838.94 sec
  
  The following tests FAILED:
  	1093 - Zoltan2_orderingTestDriverExample_MPI_1 (Failed)
```

I did an experimental submit of Zoltan2 to CDash:

* https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=3257684

The failing test `Zoltan2_orderingTestDriverExample_MPI_1` with details at:

* https://testing.sandia.gov/cdash/testDetails.php?test=43236860&build=3257685

shows the failure:

`Test driver for rank 0 caught the following exception: TPL_Traits:  Value too large for TPL index type. Rebuild TPL with larger index type or rebuild without the TPL.`

I will submit a separate GitHub issue for this failing test to investigate it further.

**Comment by mhoemmen** [comment]
@prwolfe @rrdrake

**Comment by bartlettroscoe** [comment]
@trilinos/framework,

Once this branch is merged into `develop`, and we get a few CI iterations of:

* https://testing.sandia.gov/cdash/index.php?project=Trilinos&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=66&value1=-MPI_RELEASE_DEBUG_SHARED_PT_CI&field2=groupname&compare2=61&value2=Continuous&field3=buildstarttime&compare3=84&value3=now

confirming that everything looks good, can the Trilinos Framework team please consider enabling Scotch in the builds run in automated PR testing?  Otherwise, passing automated builds in PR testing will not cover the updated set of tests when Scotch is enabled and will let the other CI build fail.

**Comment by bartlettroscoe** [comment]
This branch has been merged to `develop` (see #2054).  We should see Scotch enabled in the next CI build iteration.

**Comment by bartlettroscoe** [comment]
For some reason, adding the Scotch TPL caused configure failures in every Trilinos package when the CI server ran this morning as shown at:

* https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=3259408

The errors showed:

```
-- SEMS: Loading SEMSDevEnv.cmake to set compilers and TPL paths (To skip, set -DTrilinos_USE_BUILD_ENV=) ...
CMake Error at cmake/tribits/core/utils/AssertDefined.cmake:79 (MESSAGE):
  Error, the variable ENV{SEMS_SCOTCH_ROOT} is not defined!
Call Stack (most recent call first):
  cmake/std/sems/SEMSDevEnv.cmake:108 (ASSERT_DEFINED)
  cmake/std/sems/SEMSDevEnv.cmake:155 (SEMS_SELECT_TPL_ROOT_DIR)
  cmake/StdDevEnvs.cmake:49 (INCLUDE)
  cmake/tribits/core/package_arch/TribitsGlobalMacros.cmake:167 (INCLUDE)
  cmake/tribits/core/package_arch/TribitsProjectImpl.cmake:96 (TRIBITS_READ_IN_OPTIONS_FROM_FILE)
  cmake/tribits/core/package_arch/TribitsProject.cmake:93 (TRIBITS_PROJECT_IMPL)
  CMakeLists.txt:93 (TRIBITS_PROJECT)
```

I think this might be a result of sourcing the `load_ci_sems_dev_env.sh` script an older version of the Trilinos which did **not** load the Scotch module but then running the configure after the update expecting the Scotch module to be loaded (which it was not).

This shows the weakness in the simplistic way that that CI server was set up to only have once source repo for both the driver scripts and the inner configure.

I will restart the CI server and it should run clean this time.

**Comment by bartlettroscoe** [comment]
Comparing the initial CI build yesterday involving all packages at:

* https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=3257158

and the initial CI build today at :

* https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=3259601

you can see that the number of Zoltan tests went up from 16 to 19 and the number of Zoltan2 tests went up from 97 to 101.  Also, all of the Zoltan and Zoltan2 tests passed.

I think this is good confirmation that Scotch has been enabled and is running more tests.

@trilinos/framework,

Would you please consider adding Scotch to the set of enabled TPLs when testing with the SEMS env?

**Comment by bartlettroscoe** [comment]
I created issue #2065 to request that the @trilinos/framework team add the Scotch TPL to at least the automated PR builds.  Therefore, this story can now be closed since the CI build associated with the checkin-test-sems.sh script linked to from:

* https://github.com/trilinos/Trilinos/wiki/Policies--%7C-Testing#post_push_ci_testing

is now enabling this Scotch TPL.

Closing as complete.
